server/src/main/java/io/crate/planner/operators/WindowAgg.java

```diff
@@ -114,7 +115,7 @@ private WindowAgg(LogicalPlan source,
 
     @Override
     public LogicalPlan pruneOutputsExcept(Collection<Symbol> outputsToKeep) {
-        HashSet<Symbol> toKeep = new HashSet<>();
+        HashSet<Symbol> toKeep = new LinkedHashSet<>();
```

Comment:
Same here

Reply:
I'm really struggling to find a counter example and make it fail, e.g.:
```
@Test
public void test_prune_outputs_on_window_functions() {
    LogicalPlan plan = sqlExecutor.logicalPlan("""
        SELECT i, avgx FROM (SELECT a, avg(x) OVER(ORDER BY (SELECT t1.i)) as avgx, i FROM t1) as vt
        """);
    assertThat(plan).isEqualTo(
        """
            Rename[i, avgx] AS vt
              └ Eval[i, avg(x) OVER (ORDER BY (SELECT i FROM (empty_row)) ASC) AS avgx]
                └ WindowAgg[x, a, i, (SELECT i FROM (empty_row)), avg(x) OVER (ORDER BY (SELECT i FROM (empty_row)) ASC)]
                  └ CorrelatedJoin[x, i, (SELECT i FROM (empty_row))]
                    └ Collect[doc.t1 | [x, i] | true]
                    └ SubPlan
                      └ Eval[i]
                        └ Limit[2::bigint;0::bigint]
                          └ TableFunction[empty_row | [] | true]""");
}
```
but I cannot, don't know what to do, it feels safer to keep it as `LinkedHashSet`.

@seut what do you think?

Reply:
I slightly agree with @mfussenegger that it may not be needed in that cases as long as an `Eval` with correct ordering is applied in-between. But on the other hand, it seems that correct plans relies on the `Eval` and additionally, the overhead of using a `LinkedHashSet` in these planner phases maybe negligible?

Reply:
I agree as well, whatever I've tried so far for `ProjectSet` and `WindowsAgg`, I cannot find an example that breaks, as for those cases it is mandatory to have a subselect with an alias, which will add a `Rename` and an `Eval`, and therefore the correcting output ordering is preserved.

On the other hand, using a `LinkedHashSet` for those as well, should be a small overhead and can potentially prevent more bugs in this area in the future, so my take is to keep it, as well with those 2 additional tests, and I've already pushed a fixup to change the changelog entry and mention the one case (union+limit) which is actually fixed in this PR.

If you prefer to keep plain `HashSet` for those 2, I'm also happy to revert, as it seems there is currently no actual issue there.

Reply:
I'm fine with keeping it, so I'd leave it up to @mfussenegger.

Reply:
For me the overhead of LinkedHashSet is only a secondary factor. The main aspect is that we should understand our code enough to know if it's required.

If the `outputsToKeep` of `pruneOutputsExcept` are order sensitive and must match the original order, then the function documentation should say so and we need to use `LinkedHashSet`. (And we could mark this for upcoming https://openjdk.org/jeps/431, so the type can in the future reflect this requirement)

If it is insensitive, then it must be possible to keep using `HashSet` and we need to double check that no other operator relies on the order of `outputsToKeep`

E.g. it looks like `Eval` would currently depend on ordering of `outputsToKeep`, instead of preserving the original order (like `Collect` does)

Either way is fine, but we should have a clear contract.

Reply:
for now, I've updated the description of the method to mention keeping the order.